File: impls/es6/step9_try.mjs
import rl from './node_readline.js'
const readline = rl.readline
import { _clone, _list_Q, _malfunc, _malfunc_Q, Vector } from './types'
import { BlankException, read_str } from './reader'
import { pr_str } from './printer'
import { new_env, env_set, env_get } from './env'
import { core_ns } from './core'

// read
const READ = str => read_str(str)

// eval
const qq_loop = (acc, elt) => {
    if (_list_Q(elt) && elt.length == 2
        && elt[0] === Symbol.for('splice-unquote')) {
        return [Symbol.for('concat'), elt[1], acc]
    } else {
        return [Symbol.for('cons'), quasiquote (elt), acc]
    }
}
const quasiquote = ast => {
    if (_list_Q(ast)) {
        if (ast.length == 2 && ast[0] === Symbol.for('unquote')) {
            return ast[1]
        } else {
            return ast.reduceRight(qq_loop, [])
        }
    } else if (ast instanceof Vector) {
        return [Symbol.for('vec'), ast.reduceRight(qq_loop, [])]
    } else if (typeof ast === 'symbol' || ast instanceof Map) {
        return [Symbol.for('quote'), ast]
    } else {
        return ast
    }
}

function macroexpand(ast, env) {
    while (_list_Q(ast) && typeof ast[0] === 'symbol' && ast[0] in env) {
        let f = env_get(env, ast[0])
        if (!f.ismacro) { break }
        ast = f(...ast.slice(1))
    }
    return ast
}


const eval_ast = (ast, env) => {
    if (typeof ast === 'symbol') {
        return env_get(env, ast)
    } else if (ast instanceof Array) {
        return ast.map(x => EVAL(x, env))
    } else if (ast instanceof Map) {
        let new_hm = new Map()
        ast.forEach((v, k) => new_hm.set(EVAL(k, env), EVAL(v, env)))
        return new_hm
    } else {
        return ast
    }
}

const EVAL = (ast, env) => {
  while (true) {
    //console.log('EVAL:', pr_str(ast, true))
    if (!_list_Q(ast)) { return eval_ast(ast, env) }

    ast = macroexpand(ast, env)
    if (!_list_Q(ast)) { return eval_ast(ast, env) }
    if (ast.length === 0) { return ast }

    const [a0, a1, a2, a3] = ast
    switch (typeof a0 === 'symbol' ? Symbol.keyFor(a0) : Symbol(':default')) {
        case 'def!':
            return env_set(env, a1, EVAL(a2, env))
        case 'let*':
            let let_env = new_env(env)
            for (let i=0; i < a1.length; i+=2) {
                env_set(let_env, a1[i], EVAL(a1[i+1], let_env))
            }
            env = let_env
            ast = a2
            break // continue TCO loop
        case 'quote':
            return a1
        case 'quasiquoteexpand':
            return quasiquote(a1)
        case 'quasiquote':
            ast = quasiquote(a1)
            break // continue TCO loop
        case 'defmacro!':
            let func = _clone(EVAL(a2, env))
            func.ismacro = true
            return env_set(env, a1, func)
        case 'macroexpand':
            return macroexpand(a1, env)
        case 'try*':
            try {
                return EVAL(a1, env)
            } catch (exc) {
                if (a2 && a2[0] === Symbol.for('catch*')) {
                    if (exc instanceof Error) { exc = exc.message }
                    return EVAL(a2[2], new_env(env, [a2[1]], [exc]))
                } else {
                    throw exc
                }
            }
        case 'do':
            eval_ast(ast.slice(1,-1), env)
            ast = ast[ast.length-1]
            break // continue TCO loop
        case 'if':
            let cond = EVAL(a1, env)
            if (cond === null || cond === false) {
                ast = (typeof a3 !== 'undefined') ? a3 : null
            } else {
                ast = a2
            }
            break // continue TCO loop
        case 'fn*':
            return _malfunc((...args) => EVAL(a2, new_env(env, a1, args)),
                            a2, env, a1)
        default:
            let [f, ...args] = eval_ast(ast, env)
            if (_malfunc_Q(f)) {
                env = new_env(f.env, f.params, args)
                ast = f.ast
                break // continue TCO loop
            } else {
                return f(...args)
            }
    }
  }
}

// print
const PRINT = exp => pr_str(exp, true)

// repl
let repl_env = new_env()
const REP = str => PRINT(EVAL(READ(str), repl_env))

// core.EXT: defined using ES6
for (let [k, v] of core_ns) { env_set(repl_env, Symbol.for(k), v) }
env_set(repl_env, Symbol.for('eval'), a => EVAL(a, repl_env))
env_set(repl_env, Symbol.for('*ARGV*'), [])

// core.mal: defined using language itself
REP('(def! not (fn* (a) (if a false true)))')
REP('(def! load-file (fn* (f) (eval (read-string (str "(do " (slurp f) "\nnil)")))))')
REP('(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list \'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw "odd number of forms to cond")) (cons \'cond (rest (rest xs)))))))')

if (process.argv.length > 2) {
    env_set(repl_env, Symbol.for('*ARGV*'), process.argv.slice(3))
    REP(`(load-file "${process.argv[2]}")`)
    process.exit(0)
}


while (true) {
    let line = readline('user> ')
    if (line == null) break
    try {
        if (line) { console.log(REP(line)) }
    } catch (exc) {
        if (exc instanceof BlankException) { continue }
        if (exc instanceof Error) { console.warn(exc.stack) }
        else { console.warn(`Error: ${pr_str(exc, true)}`) }
    }
}

File: impls/cs.2/step9_try.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace mal
{
    class step9_try
    {

        static MalType READ(string input)
        {
            return Reader.read_str(input);
        }

        public static MalType EVAL(MalType ast, Env env)
        {
            while (true)
            {
                if (ast is not MalList)
                {
                    return eval_ast(ast, env);
                }
                else
                {
                    MalList astList = (MalList)ast;

                    // Remove any comments from the ast
                    List<MalType> nonComments = astList.items.Where(it => !(it is MalSymbol && ((MalSymbol)it).value.StartsWith(";"))).ToList();
                    astList = new MalList(nonComments);

                    if (astList.items.Count == 0)
                    {
                        return astList;
                    }
                    else
                    {
                        // Macroexpand before special forms
                        MalType expanded = macroexpand(astList, env);
                        if (expanded is not MalList) return eval_ast(expanded, env);
                        else
                        {
                            ast = (MalList)expanded;
                            astList = (MalList)expanded;
                        }

                        MalType first = astList.items.First();
                        if (first is MalSymbol)
                        {
                            MalSymbol firstSymbol = (MalSymbol)first;
                            if (firstSymbol.value == "def!")
                            {
                                MalSymbol symbol = (MalSymbol)astList.items[1];
                                MalType value = EVAL(astList.items[2], env);
                                env.set(symbol, value);
                                return value;
                            }
                            if (firstSymbol.value == "defmacro!")
                            {
                                MalSymbol symbol = (MalSymbol)astList.items[1];
                                MalType value = EVAL(astList.items[2], env);
                                if (value is MalFunction) ((MalFunction)value).is_macro = true;
                                if (value is MalFnTco) ((MalFnTco)value).is_macro = true;
                                env.set(symbol, value);
                                return value;
                            }
                            else if (firstSymbol.value == "let*")
                            {
                                MalSeq bindings = (MalSeq)astList.items[1];
                                MalType expression = astList.items[2];
                                Env newEnv = new Env(env);
                                for (int i = 0; i < bindings.items.Count; i += 2)
                                {
                                    MalSymbol name = (MalSymbol)bindings.items[i];
                                    MalType value = EVAL(bindings.items[i + 1], newEnv);
                                    newEnv.data.Add(name, value);
                                }
                                env = newEnv;
                                ast = expression;
                                continue;
                            }
                            else if (firstSymbol.value == "do")
                            {
                                int butLast = astList.items.Count - 2;

                                // produce side-effects and then just forget
                                List<MalType> sideEffects = astList.items.Skip(1).Take(butLast).ToList();
                                eval_ast(new MalList(sideEffects), env);

                                ast = astList.items.Last();
                                continue;
                            }
                            else if (firstSymbol.value == "if")
                            {
                                MalType test = astList.items[1];
                                MalType testResult = EVAL(test, env);
                                if (testResult == MalNil.MAL_NIL || testResult == MalBoolean.MAL_FALSE)
                                {
                                    ast = (astList.items.Count > 3) ? astList.items[3] : MalNil.MAL_NIL;
                                }
                                else
                                {
                                    ast = astList.items[2];
                                }
                                continue;
                            }
                            else if (firstSymbol.value == "fn*")
                            {
                                MalSeq argNames = (MalSeq)astList.items[1];
                                MalType funcBody = astList.items[2];
                                List<MalSymbol> argSymbs = new List<MalSymbol>();
                                foreach (MalType arg in argNames.items) { if (arg is MalSymbol) argSymbs.Add((MalSymbol)arg); }
                                MalFunction fn = new MalFunction(
                                    (IList<MalType> argValues) =>
                                    {
                                        Env funcEnv = new Env(env, argSymbs, argValues);
                                        return EVAL(funcBody, funcEnv);
                                    }
                                );
                                return new MalFnTco(funcBody, argSymbs, env, fn);
                            }
                            else if (firstSymbol.value == "quote")
                            {
                                return astList.items[1];
                            }
                            else if (firstSymbol.value == "quasiquoteexpand")
                            {
                                return quasiquote(astList.items[1]);
                            }
                            else if (firstSymbol.value == "quasiquote")
                            {
                                ast = quasiquote(astList.items[1]);
                                continue;
                            }
                            else if (firstSymbol.value == "macroexpand")
                            {
                                return macroexpand(astList.items[1], env);
                            }
                            else if (firstSymbol.value == "try*")
                            {
                                try
                                {
                                    return EVAL(astList.items[1], env);
                                }
                                catch (MalException exception)
                                {
                                    // if a catch block exists, create a MalException and evaluate the expression of the block with it
                                    if (astList.items.Count > 2 && astList.items[2] is MalList)
                                    {
                                        MalList catchArgs = (MalList)astList.items[2];
                                        MalSymbol exName = (MalSymbol)catchArgs.items[1];
                                        MalType expr = catchArgs.items[2];
                                        Env exEnv = new Env(env);
                                        exEnv.set(exName, exception.cause);
                                        return EVAL(expr, exEnv);
                                    }
                                    else throw new MalException(exception.cause); // otherwise just rethrow
                                }
                            }

                        }

                        MalType evaluated = eval_ast(ast, env);
                        if (evaluated is not MalList) return evaluated;
                        MalList evaluatedList = (MalList)evaluated;

                        // Function application
                        MalType funcFirst = evaluatedList.items[0];
                        if (funcFirst is MalFunction)
                        {
                            MalFunction func = (MalFunction)funcFirst;
                            MalType retVal = func.function(evaluatedList.items.Skip(1).ToList());
                            return retVal;
                        }
                        else // MalFnTco
                        {
                            MalFnTco fnTco = (MalFnTco)funcFirst;
                            List<MalType> fnArgs = evaluatedList.items.Skip(1).ToList();
                            Env newEnv = new Env(fnTco.env, fnTco.@params, fnArgs);
                            ast = fnTco.ast;
                            env = newEnv;
                            continue;
                        }
                    }
                }
            }
        }

        static bool is_macro_call(MalType ast, Env env)
        {
            if (ast is MalList)
            {
                MalType head = ((MalList)ast).items[0];
                if (head is MalSymbol && env.find((MalSymbol)head) != null)
                {
                    MalType value = env.get((MalSymbol)head);
                    if (value is MalFunction) return ((MalFunction)value).is_macro;
                    if (value is MalFnTco) return ((MalFnTco)value).is_macro;
                }
            }
            return false;
        }

        static MalType macroexpand(MalType ast, Env env)
        {
            bool expanding;
            while (expanding = is_macro_call(ast, env))
            {
                MalSymbol name = (MalSymbol)((MalSeq)ast).items[0];
                List<MalType> args = ((MalSeq)ast).items.Skip(1).ToList();
                MalType macroFn = env.get(name);
                if (macroFn is MalFunction) ast = ((MalFunction)macroFn).function(args);
                if (macroFn is MalFnTco) ast = ((MalFnTco)macroFn).fn.function(args);
            }
            return ast;
        }

        static MalType quasiquote(MalType ast)
        {
            if (ast is MalSeq)
            {
                MalSeq astSeq = (MalSeq)ast;
                if (ast is MalList && astSeq.items.Count > 0)
                {
                    MalType head = astSeq.items[0];
                    if (head is MalSymbol && ((MalSymbol)head).value == "unquote")
                    {
                        return astSeq.items[1];
                    }
                }

                List<MalType> result = new List<MalType>();

                // Iterate over each element elt of ast in reverse order
                for (int i = astSeq.items.Count - 1; i >= 0; i--)
                {
                    MalType elt = astSeq.items[i];

                    // If elt is a list starting with the "splice-unquote" symbol
                    if (elt is MalList && ((MalList)elt).items.Count >= 2 && ((MalList)elt).items[0].Equals(new MalSymbol("splice-unquote")))
                    {
                        // replace the current result with a list containing: the "concat" symbol,
                        // the second element of elt, then the previous result.
                        result = new List<MalType>(){
                            new MalSymbol("concat"),
                            ((MalList)elt).items[1],
                            new MalList(result)
                        };
                    }
                    else
                    {
                        // replace the current result with a list containing: the "cons" symbol,
                        // the result of calling quasiquote with elt as argument, then the previous result
                        result = new List<MalType>()
                        {
                            new MalSymbol("cons"),
                            quasiquote(elt),
                            new MalList(result)
                        };
                    }
                }

                if (ast is MalVector)
                {
                    // when ast is a vector, return a list containing: the "vec" symbol,
                    // then the result of processing ast as if it were a list not starting with quote
                    result = new List<MalType>() { new MalSymbol("vec"), new MalList(result) };
                }

                return new MalList(result);
            }
            else if (ast is MalHashmap || ast is MalSymbol)
            {
                // If ast is a map or a symbol, return a list containing: the "quote" symbol, then ast
                return new MalList(new List<MalType>() { new MalSymbol("quote"), ast });
            }
            else return ast; // Else return ast unchanged
        }

        static string PRINT(MalType input)
        {
            return printer.pr_str(input, true);
        }

        static string rep(string input)
        {
            MalType read = READ(input);
            MalType evaled;
            try
            {
                evaled = EVAL(read, repl_env);
                if (evaled == null) return null;
                string printed = PRINT(evaled);
                return printed;
            }
            catch (MalException mex)
            {
                return string.Format("Exception: {0}", printer.pr_str(mex.cause));
            }
        }

        public static MalType eval_ast(MalType ast, Env env)
        {
            if (ast is MalSymbol)
            {
                MalSymbol astSymbol = (MalSymbol)ast;
                return env.get(astSymbol);
            }
            else if (ast is MalList)
            {
                MalList astList = (MalList)ast;
                List<MalType> evaluated = astList.items.Select(item => EVAL(item, env)).ToList();
                return new MalList(evaluated);
            }
            else if (ast is MalVector)
            {
                MalVector astVector = (MalVector)ast;
                List<MalType> evaluated = astVector.items.Select(item => EVAL(item, env)).ToList();
                return new MalVector(evaluated);
            }
            else if (ast is MalHashmap)
            {
                MalHashmap astHashmap = (MalHashmap)ast;
                Dictionary<MalType, MalType> newKVs = new Dictionary<MalType, MalType>();
                foreach (KeyValuePair<MalType, MalType> kv in astHashmap.values)
                {
                    newKVs.Add(EVAL(kv.Key, env), EVAL(kv.Value, env));
                }
                return new MalHashmap(newKVs);
            }
            else
            {
                return ast;
            }
        }

        public static Env repl_env = new Env(null);

        static void Main(string[] args)
        {
            // Load the built-in functions
            foreach (var pair in core.ns) { repl_env.set(new MalSymbol(pair.Key), pair.Value); }

            // Define 'eval'
            repl_env.data.Add(new MalSymbol("eval"), new MalFunction((IList<MalType> args) =>
            {
                MalType ast = args[0];
                return EVAL(ast, repl_env);
            }));

            // Functions defined on Mal itself:
            // - define 'not'
            rep("(def! not (fn* (a) (if a false true)))");
            // - define 'load-file'
            rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\"))))))");
            rep("(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))");

            // ARGV
            List<MalType> malArgs = new List<MalType>();
            malArgs.AddRange(args.Skip(1).Select(arg => new MalString(arg)).ToList());
            repl_env.data.Add(new MalSymbol("*ARGV*"), new MalList(malArgs));

            // if called with arguments, treat first as a script name
            if (args.Length > 0)
            {
                rep("(load-file \"" + args[0] + "\")");
                return;
            }

            // TESTS
            // rep("`[unquote 0]");

            string line = null;
            do
            {
                Console.Write("user> ");
                line = Console.ReadLine();
                if (line != null)
                {
                    try
                    {
                        string repResult = rep(line);
                        if (repResult != null) Console.WriteLine(repResult);
                    }
                    catch (Exception ex)
                    {

                        Console.WriteLine(ex.Message);
                    }
                }
            } while (line != null);
            Console.WriteLine();
        }
    }
}
